Source/CarthageKitTests/Fakes/FakeURLSession.swift

```diff
@@ -0,0 +1,41 @@
+import Foundation
+
+final class FakeDownloadTask: NSURLSessionDownloadTask {
+	var started = false
+	override func resume() {
+		self.started = true
+	}
+
+	var cancelled = false
+	override func cancel() {
+		self.cancelled = true
+	}
+}
+
+final class FakeDataTask: NSURLSessionDataTask {
+	var started = false
+	override func resume() {
+		self.started = true
+	}
+
+	var cancelled = false
+	override func cancel() {
+		self.cancelled = true
+	}
+}
+
+final class FakeURLSession: NSURLSession {
```

Comment:
> I definitely agree about the code as it stands today. It really lacks good structure. I think that's the problem: that we have the wrong pure functions, not that we have too many of them. But that's just intuition and could easily be wrong.

I think it was more that it started as something good, but has since grown unchecked and hasn't been pruned back.

> And if you think the integration tests as you have them in this PR would be helpful in refactoring and cleaning things up so we can better test, then I will happily merge them! If nothing else, they might be a good starting place for a PR to cleanup the code: add the tests, refactor the code, unit test the code, then remove some or all of the integration tests.

Hmm, this is somewhat troubling. I'd like to move to something that can be expressed as `handleNetworkResults(fetchFromNetwork(GitHubRequest(githubRepo, authenticationParam)))` , but that's not really within the scope of this pull request.

In the meantime, though, I think it would be better to have this area of code be tested, but acknowledge that we want to change it.

Bottom line, yes, the current tests will help as it gets refactored, so please merge this. I'll open up a separate pull request (maybe as soon as later today) refactor this again.